dss/storage/validator.py

```diff
@@ -1,23 +1,91 @@
+import io
 import re
+
+import boto3
+from cloud_blobstore import BlobNotFoundError
 from jsonschema import validators
 from jsonschema import _utils, _validators
 
 import json
-from urllib.request import urlopen
-from urllib.error import URLError
-import ssl
-from typing import List
+from typing import List, Optional
 import logging
 
+from requests import RequestException
+
+from dss import Replica, Config, DSSException
+from hashlib import sha1
+import requests
+
+
+DSS_Draft4Validator = validators.create(meta_schema=_utils.load_schema("draft4"),
+                                        validators={u"$ref": _validators.ref,
+                                                    u"additionalProperties": _validators.additionalProperties,
+                                                    u"properties": _validators.properties_draft4,
+                                                    u"required": _validators.required_draft4
+                                                    },
+                                        version="draft4"
+                                        )
+
+
+class SizeLimitError(requests.RequestException):
+    "The max size content size has been exceeded"
+    pass
+
+class S3UrlCache:
+    """Caches content of arbitrary URLs the first time they are requested. Currently only supports content lengths of up
+     to a few megabytes."""
+
+    def __init__(self, logger: logging.Logger):
+        self.s3_client = boto3.client('s3')
+
+        # A URL's contents are only stored in s3 to keep the data closer to the aws lambas which use them.
+        self.blobstore, _, self.bucket = Config.get_cloud_specific_handles(Replica.aws)
+        self.logger = logger
 
-def read_json_url(link):
-    gcontext = ssl.SSLContext(ssl.PROTOCOL_TLSv1)
-    with urlopen(link, context=gcontext) as f:
-        return json.load(f)
+    def request(self, url: str, max_size: Optional[int]=64000) -> str:
+        """
+        Requests the contents of a URL by first checking for the contents in an s3 bucket. If not present, the contents
+        are retrieved from the URL and stored in s3.
 
+        :param url: The JSON data from which to remove fields.
+        :param size: The max number of bytes to retrieve from URL.
+        """
+        key = 'cache/' + sha1(url.encode("utf-8")).hexdigest()
 
-def remove_json_fields(json_data: dict, path: List[str], fields: List[str]):
+        # if key in s3 bucket return value from there.
+        try:
+            content = self.blobstore.get(self.bucket, key)
+        except BlobNotFoundError:
+            self.logger.info("%s", f"{url} not found in cache. Adding it to {self.bucket} with key {key}.")
+            with requests.get(url, stream=True) as f:
+                content = b''
+                for chunk in f.iter_content(chunk_size=None):
+                    content = content + chunk
+                    if len(content) > max_size:
```

Comment:
changed to:
```python
                for chunk in resp_obj.iter_content(chunk_size=None):
                    #  check if max_size exceeded before storing content to avoid storing large chunks
                    if len(content) + len(chunk) > self.max_size:
                        raise SizeLimitError("{url} not cached. Contents have exceed {self.max_size} Bytes.")
                    content.extend(chunk)
```